photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/search/GaussianProcessSearch.scala

```diff
@@ -0,0 +1,167 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.hyperparameter.search
+
+import math.max
+
+import breeze.linalg.{DenseMatrix, DenseVector}
+import breeze.numerics.sqrt
+import breeze.stats.variance
+import breeze.stats.distributions.{RandBasis, ThreadLocalRandomGenerator, Multinomial, Uniform}
+import org.apache.commons.math3.random.MersenneTwister
+
+import com.linkedin.photon.ml.evaluation.Evaluator
+import com.linkedin.photon.ml.hyperparameter.EvaluationFunction
+import com.linkedin.photon.ml.hyperparameter.criteria.{ConfidenceBound, ExpectedImprovement}
+import com.linkedin.photon.ml.hyperparameter.estimators.{GaussianProcessEstimator, GaussianProcessModel, PredictionTransformation}
+import com.linkedin.photon.ml.hyperparameter.estimators.kernels.Matern52
+
+/**
+ * Performs a guided random search of the given ranges, where the search is guided by a Gaussian Process estimated from
+ * evaluations of the actual evaluation function. Since we assume that the evaluation function is very costly (as it
+ * often is for doing a full train / cycle evaluation of a machine learning model), it makes sense to spend time doing
+ * what would otherwise be considered an expensive computation to reduce the number of times we need to evaluate the
+ * function.
+ *
+ * At a high level, the search routine proceeds as follows:
+ *
+ *  1) Assume a uniform prior over the evaluation function
+ *  2) Receive a new observation, and use it along with any previous observations to train a new Gaussian Process
+ *     regression model for the evaluation function. This approximation is the new posterior over the evaluation
+ *     function.
+ *  3) Sample candidates uniformly, evaluate the posterior for each, and select the candidate with the highest predicted
+ *     evaluation.
+ *  4) Evaluate the best candidate with the actual evaluation function to acquire a new observation.
+ *  5) Repeat from step 2.
+ *
+ * @param ranges the value ranges within which to search. There should be one for every dimension in the space.
+ * @param evaluationFunction the function that evaluates points in the space to real values
+ * @param evaluator the original evaluator
+ * @param candidatePoolSize the number of candidate points to draw at each iteration. Larger numbers give more precise
+ *   results, but also incur higher computational cost.
+ * @param seed the random seed value
+ */
+class GaussianProcessSearch[T](
+    ranges: Seq[(Double, Double)],
+    evaluationFunction: EvaluationFunction[T],
+    evaluator: Evaluator,
+    candidatePoolSize: Int = 250,
+    seed: Long = System.currentTimeMillis)
+  extends RandomSearch[T](ranges, evaluationFunction, seed){
```

Comment:
I'd considered this, but the GaussianProcessSearch is technically a random search, though it is guided. I agree that this relationship is a little awkward, though. Going to give this a bit more thought.